### Import required dependencies

In [392]:
import pandas as pd
import os

## Deliverable 1: Collect the Data

To collect the data that you’ll need, complete the following steps:

1. Using the Pandas `read_csv` function and the `os` module, import the data from the `new_full_student_data.csv` file, and create a DataFrame called student_df. 

2. Use the head function to confirm that Pandas properly imported the data.


In [393]:
# Create the path and import the data
full_student_data = os.path.join('../Resources/new_full_student_data.csv')
student_df = pd.read_csv(full_student_data)

In [394]:
# Verify that the data was properly imported
student_df.head()

student_id      student_name grade           school_name  reading_score  \
0   103880842     Travis Martin   9th  Sullivan High School           59.0   
1    45069750     Michael Brown   9th     Dixon High School           94.7   
2    45024902   Gabriela Lucero   9th    Wagner High School           89.0   
3    62582498  Susan Richardson   9th     Silva High School           69.7   
4    16437227      Sherry Davis  11th    Bowers High School            NaN   

   math_score school_type  school_budget  
0        88.2      Public         961125  
1        73.5     Charter         870334  
2        70.4      Public         846745  
3        80.3      Public         991918  
4        27.5      Public         848324

## Deliverable 2: Prepare the Data

To prepare and clean your data for analysis, complete the following steps:
    
1. Check for and remove all rows with `NaN`, or missing, values in the student DataFrame. 

2. Check for and remove all duplicate rows in the student DataFrame.

3. Use the `str.replace` function to remove the "th" from the grade levels in the grade column.

4. Check data types using the `dtypes` property.

5. Remove the "th" suffix from every value in the grade column using `str` and `replace`.

6. Change the grade colum to the `int` type and verify column types.

7. Use the head (and/or the tail) function to preview the DataFrame.

In [395]:
# Check for null values
student_df.count()

student_id       19514
student_name     19514
grade            19514
school_name      19514
reading_score    17546
math_score       18532
school_type      19514
school_budget    19514
dtype: int64

In [396]:
# Drop rows with null values and verify removal
new_student_df = student_df.dropna(how='any')

In [397]:
# verify removal
new_student_df.count()

student_id       16667
student_name     16667
grade            16667
school_name      16667
reading_score    16667
math_score       16667
school_type      16667
school_budget    16667
dtype: int64

In [398]:
# Check for duplicated rows
new_student_df.duplicated()

0        False
1        False
2        False
3        False
5        False
         ...  
19509    False
19510     True
19511    False
19512    False
19513    False
Length: 16667, dtype: bool

In [399]:
# Drop duplicated rows and verify removal
new_student_df = new_student_df.drop_duplicates()

In [400]:
# Verify removal
new_student_df.duplicated()

0        False
1        False
2        False
3        False
5        False
         ...  
19508    False
19509    False
19511    False
19512    False
19513    False
Length: 14831, dtype: bool

In [401]:
# Check data types
new_student_df.dtypes

student_id         int64
student_name      object
grade             object
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

In [402]:
# Examine the grade column to understand why it is not an int
new_student_df.head()

student_id      student_name grade             school_name  reading_score  \
0   103880842     Travis Martin   9th    Sullivan High School           59.0   
1    45069750     Michael Brown   9th       Dixon High School           94.7   
2    45024902   Gabriela Lucero   9th      Wagner High School           89.0   
3    62582498  Susan Richardson   9th       Silva High School           69.7   
5    74579444   Cynthia Johnson   9th  Montgomery High School           63.5   

   math_score school_type  school_budget  
0        88.2      Public         961125  
1        73.5     Charter         870334  
2        70.4      Public         846745  
3        80.3      Public         991918  
5        76.9     Charter         893368

In [403]:
# Remove the non-numeric characters and verify the contents of the column
new_student_df['grade'] = new_student_df['grade'].str.replace('th', '')
new_student_df['grade']

0         9
1         9
2         9
3         9
5         9
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 14831, dtype: object

In [404]:
# Change the grade column to the int type and verify column types
new_student_df['grade'] = new_student_df['grade'].astype('int64')
new_student_df.dtypes

student_id         int64
student_name      object
grade              int64
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

## Deliverable 3: Summarize the Data

Describe the data using summary statistics on the data as a whole and on individual columns.

1. Generate the summary statistics for each DataFrame by using the `describe` function.

2. Display the mean math score using the `mean` function. 

2. Store the minimum reading score as `min_reading_score`.

In [405]:
# Display summary statistics for the DataFrame
new_student_df.describe()

student_id         grade  reading_score    math_score  school_budget
count  1.483100e+04  14831.000000   14831.000000  14831.000000   14831.000000
mean   6.975296e+07     10.355539      72.357865     64.675733  893742.749107
std    3.452909e+07      1.097728      15.224590     15.844093   53938.066467
min    1.000906e+07      9.000000      10.500000      3.700000  817615.000000
25%    3.984433e+07      9.000000      62.200000     54.500000  846745.000000
50%    6.965978e+07     10.000000      73.800000     65.300000  893368.000000
75%    9.927449e+07     11.000000      84.000000     76.000000  956438.000000
max    1.299997e+08     12.000000     100.000000    100.000000  991918.000000

In [406]:
# Display the mean math score using the mean function
mean = new_student_df["math_score"].mean()
mean

64.67573326141189

In [407]:
# Store the minimum reading score as min_reading_score
min_reading_score = new_student_df["reading_score"].min()
min_reading_score

10.5

## Deliverable 4: Drill Down into the Data

Drill down to specific rows, columns, and subsets of the data.

To drill down into the data, complete the following steps:

1. Use `loc` to display the grade column.

2. Use `iloc` to display the first 3 rows and columns 3, 4, and 5.

3. Show the rows for grade nine using `loc`.

4. Store the row with the minimum overall reading score as `min_reading_row` using `loc` and the `min_reading_score` found in Deliverable 3.

5. Find the reading scores for the school and grade from the output of step three using `loc` with multiple conditional statements.

6. Using conditional statements and `loc` or `iloc`, find the mean reading score for all students in grades 11 and 12 combined.

In [408]:
# Use loc to display the grade column
new_student_df.loc[:,["grade"]]

grade
0          9
1          9
2          9
3          9
5          9
...      ...
19508     10
19509     12
19511     11
19512     11
19513     12

[14831 rows x 1 columns]

In [409]:
# Use `iloc` to display the first 3 rows and columns 3, 4, and 5.
new_student_df.iloc[0:3, [3,4,5]]

school_name  reading_score  math_score
0  Sullivan High School           59.0        88.2
1     Dixon High School           94.7        73.5
2    Wagner High School           89.0        70.4

In [410]:
# Select the rows for grade nine and display their summary statistics using `loc` and `describe`.
new_student_df.loc[new_student_df["grade"] == 9].describe()


student_id   grade  reading_score   math_score  school_budget
count  4.132000e+03  4132.0    4132.000000  4132.000000    4132.000000
mean   6.979441e+07     9.0      69.236713    66.585624  898692.606002
std    3.470565e+07     0.0      15.277354    16.661533   54891.596611
min    1.000906e+07     9.0      17.900000     5.300000  817615.000000
25%    3.953848e+07     9.0      59.000000    56.000000  846745.000000
50%    6.984037e+07     9.0      70.050000    67.800000  893368.000000
75%    9.939504e+07     9.0      80.500000    78.500000  957299.000000
max    1.299997e+08     9.0      99.900000   100.000000  991918.000000

In [411]:
# Store the row with the minimum overall reading score as `min_reading_row`
min_reading_score = new_student_df["reading_score"].min()
min_reading_score_row = new_student_df.loc[new_student_df["reading_score"] == min_reading_score]
min_reading_score_row


student_id    student_name  grade        school_name  reading_score  \
3706    81758630  Matthew Thomas     10  Dixon High School           10.5   

      math_score school_type  school_budget  
3706        58.4     Charter         870334

In [412]:
# Use loc with conditionals to select all reading scores from 10th graders at Dixon High School.
#Filter out by grade 10
grade_ten_df = new_student_df.loc[new_student_df["grade"] == 10, :]
grade_ten_df.loc[grade_ten_df["school_name"] == "Dixon High School", ["school_name", "reading_score"]]


school_name  reading_score
45     Dixon High School           71.1
60     Dixon High School           59.5
69     Dixon High School           88.6
94     Dixon High School           81.5
100    Dixon High School           95.3
...                  ...            ...
19283  Dixon High School           52.9
19306  Dixon High School           58.0
19344  Dixon High School           38.0
19368  Dixon High School           84.4
19445  Dixon High School           43.9

[569 rows x 2 columns]

In [413]:
# Find the mean reading score for all students in grades 11 and 12 combined.
grade_eleven_and_twelve_df = new_student_df.loc[new_student_df["grade"] > 10, :]
eleven_and_twelve_mean = grade_eleven_and_twelve_df["reading_score"].mean()
eleven_and_twelve_mean

74.90038089192188

In [414]:
grade_ten_df = new_student_df["grade"] == 10
new_student_df.loc[grade_ten_df]

student_id     student_name  grade             school_name  \
8        75984300   Victoria Lewis     10    Campbell High School   
11      105535412      Ashley Ryan     10       Green High School   
19       62962791     Taylor Nixon     10       Green High School   
20       68272498    Arthur Zuniga     10      Campos High School   
21      118221043        Alex Cain     10       Green High School   
...           ...              ...    ...                     ...   
19490    85748028        Troy Howe     10      Fisher High School   
19496    19588063        Kim Moore     10    Odonnell High School   
19504    26517797     Sierra Grant     10       Green High School   
19505    56411342     Brandon Beck     10  Montgomery High School   
19508    83985333  Deborah Sanders     10       Silva High School   

       reading_score  math_score school_type  school_budget  
8               65.7        96.1     Charter         960726  
11              80.0        74.8     Charter         832670  
19              77.5        48.1     Charter         832670  
20              71.6        91.3      Public         957299  
21              64.3        60.3     Charter         832670  
...              ...         ...         ...            ...  
19490           97.6        44.1      Public         817615  
19496           65.3        53.1      Public         959474  
19504           67.6        64.2     Charter         832670  
19505           66.6        43.0     Charter         893368  
19508           60.5        64.6      Public         991918  

[4398 rows x 8 columns]

## Deliverable 5: Make Comparisons Between District and Charter Schools

Compare district vs charter schools for budget, size, and scores.

Make comparisons within your data by completing the following steps:

1. Using the `groupby` and `mean` functions, look at the average budget per school type.

1. Using the `groupby` and `count` functions, find the total number of students at each school in descending order.

2. Using the `groupby` and `mean` functions, find the average math score by grade for each school type.

In [415]:
# Display the average budget for each school type by using the groupby and mean functions
school_type_group_df = new_student_df.groupby("school_type").mean()
school_type_group_df.loc[:, ["school_budget"]]

school_budget
school_type               
Charter      872625.656236
Public       911195.558251

In [416]:
# Use the `groupby`, `count`, and `sort_values` functions to find the
# total number of students at each school and sort from most students to least students.
school_type_group_df = new_student_df.groupby("school_name").count()
school_type_group_df.loc[:, ["student_name"]].sort_values("student_name", ascending=False)

student_name
school_name                         
Montgomery High School          2038
Green High School               1961
Dixon High School               1583
Wagner High School              1541
Silva High School               1109
Woods High School               1052
Sullivan High School             971
Turner High School               846
Bowers High School               803
Fisher High School               798
Richard High School              551
Campos High School               541
Odonnell High School             459
Campbell High School             407
Chang High School                171

In [433]:
#Find the average math score by grade for each school type by using the groupby and mean functions
school_type_group_df = new_student_df.groupby(["school_type", "grade"])
# Find mean and round to one decimal place. 
# Found code for rounding to certain decimal at (https://stackoverflow.com/questions/45451189/in-pandas-can-you-aggregate-by-mean-and-round-that-mean-to-the-nearest-int)
school_type_group_df = school_type_group_df[["math_score"]].mean().round()
school_type_group_df


math_score
school_type grade            
Charter     9            70.0
            10           66.0
            11           68.0
            12           60.0
Public      9            64.0
            10           64.0
            11           59.0
            12           64.0

# Deliverable 6: Summarize Your Findings
In the cell below, write a few sentences to describe any discoveries you made while performing your analysis along with any additional analysis you believe would be worthwhile.

The first discovery we had involved comparing the budgets of Charter schools vs. Public schools. We can look at the data and see that Public schools had an higher budget of about $4,000

The next discovery we saw involved the data where we sorted the schools by student count and listed them in descending order. There was quite a difference in student count when looking at the most populated (2,038 students) and the least populated (171 students). Nine of the fifteen schools we looked at all had less than one thousand students

The last discovery that we found while looking at the drilled down data involved looking at the math scores amongst the grades in Charter and Public schools. While looking at the Charter schools, we see that 9th graders had the highest math score when compared to the other grades. There was a difference of almost 10 points when comparing 9th and 12th grade scores. When we compared those results to the public schools, there was a lot less variance among the grade levels. The scores were slightly lower when compared the Charter schools, but there was not as much of a swing like we saw in the Charter schools

I think that something else that would be interesting to look at would be to add in the reading score column with the math scores in the last exercise. That way we could compare the two different scores because it's possible that one set of schools focused on reading more than math scores. That would let the school know that they may need to spend a little less time on reading scores and more time on math scores to get the two closer to one another. The same could go for the grade level where there may be a tradeoff in certain grades where one of the scores is greatly different than the other which would mean more resources should go to bringing up the lower score of the two.